### PR TITLE
cksum: Accept non UTF-8 inputs

### LINKED ISSUE #6573: cksum: --check can't handle non-UTF-8 in comments
Currently, we convert every line to a `String`, which has the disadvantage of not properly handling malformed UTF-8, even if it appears in an "irrelevant" place:

```console
$ cat <<EOF | sed -re 's,xxx,\xff,' > sums
SHA256 (README.md) = 4d9b4c7b50b976ec5b0b2dce5a767d2da4b63eb68592017d475c4e11799a333c
# Comment shouldn't matter: >>xxx<<
SHA256 (LICENSE) = 29953405eaa3dcc41c37d1621d55b6a47eee93e05613e439e73295029740b10c
EOF
$ cksum --check sums
README.md: OK
LICENSE: OK
$ cargo run -q cksum --check sums
cksum: stream did not contain valid UTF-8
[$? = 1]
```

It also isn't supported in other places during `check`:

```console
$ echo 'U0hBMjU2IChmdW5ref9uYW1lKSA9IDI5OTUzNDA1ZWFhM2RjYzQxYzM3ZDE2MjFkNTViNmE0N2VlZTkzZTA1NjEzZTQzOWU3MzI5NTAyOTc0MGIxMGMK' | base64 -d > sums
$ cksum -c sums 
cksum: 'funky'$'\377''name': No such file or directory
funky�name: FAILED open or read
cksum: WARNING: 1 listed file could not be read
[$? = 1]
$ cargo run -q cksum -c sums 
cksum: stream did not contain valid UTF-8
[$? = 1]
```

### PR DESCRIPTION
**Pull request by RenjiSann**
Fixes #6573

### CODE REVIEW

**Comment by RenjiSann** [comment]
@BenWiederhake Could you take a look :)

File: src/uucore/src/lib/features/checksum.rs [hunk 1]
```diff
@@ -308,7 +311,21 @@ fn determine_regex(lines: &[String]) -> Option<(Regex, bool)> {
...
+            if !first.is_ascii_whitespace() {
+                break;
+            }
+            line_trim = &line_trim[1..];
+        }
+        while let Some(last) = line_trim.last() {
+            if !last.is_ascii_whitespace() {
+                break;
+            }
+            line_trim = &line_trim[..line_trim.len() - 1];
```
Comment (sylvestre) [inline src/uucore/src/lib/features/checksum.rs:326]: could you please document what this line is doing, it isn't obvious

File: src/uucore/src/lib/features/checksum.rs [hunk 2]
```diff
@@ -349,16 +362,21 @@ fn get_expected_checksum(
             )),
         }
     } else {
-        Ok(caps.name("checksum").unwrap().as_str().to_string())
+        Ok(
+            std::str::from_utf8(caps.name("checksum").unwrap().as_bytes())
```
Comment (sylvestre) [inline src/uucore/src/lib/features/checksum.rs:366]: please just use from_utf8 and import it

File: src/uucore/src/lib/features/checksum.rs [hunk 3]
```diff
@@ -488,8 +514,32 @@ where
             }
         };
 
-        let reader = BufReader::new(file);
-        let lines: Vec<String> = reader.lines().collect::<Result<_, _>>()?;
+        let mut reader = BufReader::new(file);
+        let lines = iter::from_fn(|| {
```
Comment (sylvestre) [inline src/uucore/src/lib/features/checksum.rs:518]: maybe move this into a function?

File: tests/by-util/test_cksum.rs [hunk 4]
```diff
@@ -1277,3 +1277,27 @@ fn test_non_utf8_filename() {
         .stdout_is_bytes(b"SHA256 (funky\xffname) = e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\n")
         .no_stderr();
 }
+
+#[cfg(target_os = "linux")]
```
Comment (sylvestre) [inline tests/by-util/test_cksum.rs:1281]: why linux only ?

could you please to add a test to inject invalid utf8 ?

Comment (RenjiSann) [reply tests/by-util/test_cksum.rs:1281]: > why linux only ?

This is somewhat related to [this](https://github.com/uutils/coreutils/pull/6563#issuecomment-2225260360).

TLDR: With the current MSRV being 1.70, on windows. there is no way to translate between `OsStr` and `&[u8]` without going through `str` (this holds for owned types as well).

> could you please to add a test to inject invalid utf8 ?

This is what the test should be doing. Using non UTF-8 characters that would make the old implementation fail because it would try to create `String`s from these characters.

File: src/uucore/src/lib/features/checksum.rs (hunk 2 above)
Comment (RenjiSann) [reply src/uucore/src/lib/features/checksum.rs:366]: Done :+1:

File: src/uucore/src/lib/features/checksum.rs (hunk 1 above)
Comment (RenjiSann) [reply src/uucore/src/lib/features/checksum.rs:326]: Will do.

It is a simple implementation of the `trim_ascii()` function which is not available for MSRV 1.70.

File: src/uucore/src/lib/features/checksum.rs (hunk 3 above)
Comment (RenjiSann) [reply src/uucore/src/lib/features/checksum.rs:518]: I've done it in #6654, if you're OK with letting it like this here I'll be happy, cause I'd prefer to avoid messy rebasing is possible.

**Comment by RenjiSann** [comment]
@sylvestre all your comments are addressed in #6654. If that's still an issue, please tell me, I will refactor the code accordingly.

File: tests/by-util/test_cksum.rs (hunk 4 above)
Comment (BenWiederhake) [reply tests/by-util/test_cksum.rs:1281]: > there is no way to translate between OsStr and &[u8]

Yes there is: `os_str_as_bytes`, just like you also use in the crate itself.

> This is what the test should be doing.

No, the test does not check for correct handling of *files* with non-UTF-8 names.

**Review by BenWiederhake — CHANGES_REQUESTED** [review]
I'm confused.

- This PR has older code, doesn't address the points sylvestre already mentioned, and you also say that #6654 should be used instead.
- The other PR (#6654) doesn't address sylvestre's points either (e.g. the linux-only issue), and seems to be a collection of many different PRs, which makes it unnecessarily difficult to review, and it's also in draft mode (which indicates to me that it's more of a "sneak peek", and not ready).

So which PR would you like us to review?

File: tests/by-util/test_cksum.rs (hunk 4 above)
Comment (RenjiSann) [reply tests/by-util/test_cksum.rs:1281]: I'd like to be sure I get it correctly. In the added test, I should check that the `hashes` variable get written in a file with a non-UTF-8 name, is this correct ?

If yes, I might need to refactor slightly the test framework in order to be able to write to files with non-UTF-8 names.

**Comment by RenjiSann** [comment]
> The other PR doesn't address sylvestre's points either (e.g. the linux-only issue)

The test is linux-only because `os_str_as_bytes` will automatically fail if given non-UTF-8 characters on windows.
This is due to the fact that there is no safe provided way to convert a `OsStr` to `&[u8]` on windows, and we are required to go through an intermediary `&str` conversion.

Improving the handling of windows filesystem exceptions this might be a further enhancement, but I don't think this should be the main point of this PR.

> and seems to be a collection of many different PRs, which makes it unnecessarily difficult to review,

True, #6654 is a really big refactor, which solves many problems at once. I think about re-organizing all the changes in smaller PRs, while trying to keep the reasoning clear.

File: src/uucore/src/lib/features/checksum.rs (hunk 2 above)
Comment (RenjiSann) [reply src/uucore/src/lib/features/checksum.rs]: I moved the change to this PR.

File: src/uucore/src/lib/features/checksum.rs (hunk 3 above)
Comment (RenjiSann) [reply src/uucore/src/lib/features/checksum.rs]: I moved the change into this PR

**Comment by RenjiSann** [comment]
As requested, I changed the code to handle non-UTF8 filenames, and I added a test for it.

Previously, I used `String::from_utf8_lossy` to display the filename on the terminal when needed. I realized this was wrong, because it actually inserts a REPLACEMENT CHARACTER `U+FFFD` for any non-UTF-8 sequence, when we actually want to omit them completely.
At first, I tried to copy-paste the implementation of `String::from_utf8_lossy` and adapt it, but I figured out, because of unstabilized internals, that this could not be done without copy-pasting even more code from the stdlib.

The compromise I went with is to just remove the `U+FFFD` chars from the output of `String::from_utf8_lossy`. It might produce unwanted behavior in case the character appears in the original filename, but I consider this to be unlikely enough to wait to fix it when the involved `String`'s API is stable (MSRV 1.79 iirc).

**Comment by RenjiSann** [comment]
@BenWiederhake If you want to take a look, this should finally be mergeable

File: tests/by-util/test_cksum.rs [hunk 5]
```diff
@@ -1277,3 +1277,50 @@ fn test_non_utf8_filename() {
...
+
+    at.touch(&filename);
+    at.write_bytes("check", b"SHA256 (funky\xffname) = e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\n");
+
+    scene
+        .ucmd()
+        .arg("--check")
+        .arg(at.subdir.join("check"))
+        .succeeds()
+        .stdout_is("funkyname: OK\n")
```
Comment (BenWiederhake) [inline tests/by-util/test_cksum.rs:1324]: This is the wrong output. If the filename contains `\xff`, then the output should contain *something* in its place.

- If you decide it's in-scope for this PR, then please change the code to emit something like that.
- If you decide it's out-of-scope for this PR, then please at least add a FIXME here plus another FIXME in the code where the filename is emitted, to indicate that the code is technically incorrect.

File: src/uucore/src/lib/features/checksum.rs [hunk 6]
```diff
@@ -565,18 +607,18 @@ where
...
                     }
                     res.failed_cksum += 1;
                 }
             } else {
-                if line.is_empty() {
-                    // Don't show any warning for empty lines
+                if line.is_empty() || line_bytes.starts_with(b"#") {
+                    // Don't show any warning for empty or commented lines
                     continue;
                 }
```
Comment (BenWiederhake) [inline src/uucore/src/lib/features/checksum.rs:623]: This adds support for commented-out lines. Please don't sneakily implement features like this by hiding it inside a bigger changeset. Also, please add tests to check that this actually works as intended, to prevent future regressions.

**Review by BenWiederhake — CHANGES_REQUESTED** [review]
- Please add a platform-independent test that checks the comment-feature
- Please decide on how to handle the "incorrect" output. It's fine if you don't want to deal with it and just want to get this merged, but please at least leave behind a FIXME to mark the places where the code is still technically incorrect.

File: src/uucore/src/lib/features/checksum.rs (hunk 6 above)
Comment (RenjiSann) [reply src/uucore/src/lib/features/checksum.rs:623]: Ok, I am moving this into its own PR

Comment (RenjiSann) [reply src/uucore/src/lib/features/checksum.rs:623]: Check #6782

File: tests/by-util/test_cksum.rs (hunk 5 above)
Comment (RenjiSann) [reply tests/by-util/test_cksum.rs:1324]: You are right, I confused the outputs when doing my tests

**Comment by RenjiSann** [comment]
I closing this PR in favor of these 2 new ones : 
- comments handling (#6782)
- non-UTF-8 handling (#6793)
